### PR TITLE
Refactor jubjub and shares (stability improvements)

### PR DESCRIPTION
**Pull request by Drake-Eidukas**
This refactors jubjub to explicitly use `ShareFuture`s, which helps clean up its interface and usage greatly. This accomplishes this my making its methods no longer async, but adding in `open` and `equals`, which are async and provide boilerplate code that was otherwise needed. This also updates how we assign ids to incoming shares (there were some issues with it).

Finally, this updates the existing code to use `create_task` instead of `ensure_future`, and updates existing jubjub code (mimc, etc.)

### CODE REVIEW

**Comment by Drake-Eidukas** [comment]
Though this seemed to have a speedup for `prog` tests, overall, the impact is only around 10s total across all the unit tests when running with `-n0`. With `-nauto`, sometimes I see pretty good speedups, but that has more to do with the execution order of tests more than anything else (our fastest tests run first, which limits the utility of parallelizing our tests).

**Review by hanyunx — APPROVED** [review]
Looks good to me!

File: honeybadgermpc/mpc.py [hunk 1]
```diff
@@ -64,11 +64,27 @@ def __init__(self, sid, n, t, myid, pid, send, recv, prog, config, **prog_args):
...
+            kwargs(dict): named arguments to pass to the call to the mixin
+
+        outputs:
+            future that resolves to the result of calling the mixin operation
+        """
         if name not in self.config:
             raise NotImplementedError(f"Mixin {name} not present!")
 
-        return asyncio.ensure_future(self.config[name](self, *args, **kwargs))
+        return asyncio.create_task(self.config[name](self, *args, **kwargs))
```
Comment (smkuls) [inline honeybadgermpc/mpc.py:89]: why one over the other?

File: honeybadgermpc/mpc.py [hunk 2]
```diff
@@ -64,11 +64,27 @@ def __init__(self, sid, n, t, myid, pid, send, recv, prog, config, **prog_args):
         self.GFElementFuture = type(
             'GFElementFuture', (GFElementFuture,), {'context': self})
 
+    def _get_share_id(self):
```
Comment (smkuls) [inline honeybadgermpc/mpc.py:68]: Ummmm, there are some serious issues in the way we assign the `ShareId`. @rahulgovind and I had discussed an example in which assigning the order this way won't be consistent across all parties since the tasks are not executed in a deterministic order.

Why did you see a need to change this?

File: honeybadgermpc/mpc.py [hunk 3]
```diff
@@ -92,47 +108,65 @@ def call_mixin(self, name, *args, **kwargs):
...
+            self.send(dest, ('S', shareid, value_to_share))
 
         # Set up the buffer of received shares
         share_buffer = [self._share_buffers[i][shareid] for i in range(self.N)]
 
         point = EvalPoint(self.field, self.N, use_fft=False)
 
         # Create polynomial that reconstructs the shared value by evaluating at 0
-        opening = robust_reconstruct(
+        p, _ = await robust_reconstruct(
```
Comment (smkuls) [inline honeybadgermpc/mpc.py]: what's the second return value? Why are we ignoring it? Should we log it?

File: honeybadgermpc/mpc.py [hunk 4]
```diff
@@ -92,47 +108,65 @@ def call_mixin(self, name, *args, **kwargs):
...
             share_buffer, self.field, self.N, t, point)
-        self._openings[shareid] = opening
-
-        p, _ = await opening
 
         # Return reconstructed point
         return p(self.field(0))
 
-    def open_share_array(self, sharearray):
+    async def open_share_array(self, sharearray):
```
Comment (smkuls) [inline honeybadgermpc/mpc.py]: Why is this change required?

File: honeybadgermpc/mpc.py [hunk 5]
```diff
@@ -92,47 +108,65 @@ def call_mixin(self, name, *args, **kwargs):
...
+        # bgtask should not exit early-- this should correspond to an error
+        if bgtask.done():
+            logging.error('Background task finished before prog')
+
+            bg_exception = bgtask.exception()
+            if not result.done():
+                result.cancel()
+
+            if bg_exception is None:
+                raise Exception('background task finished before prog!')
```
Comment (smkuls) [inline honeybadgermpc/mpc.py]: `raise HoneyBadgerMpcException`, avoid raising generic exceptions!

Can you think of a case when this can happen?
As a rule, always log exceptions, we have seen things hang when there are exceptions but it is easier to spot a log.

File: honeybadgermpc/progs/jubjub.py [hunk 6]
```diff
@@ -198,15 +194,24 @@ def __str__(self):
...
         if not isinstance(n, int):
             raise Exception("Can't scale a point by something which isn't an int!")
 
         return self
 
-    async def double(self):
+    def double(self):
         return self
 
+    async def equals(self, other):
```
Comment (smkuls) [inline honeybadgermpc/progs/jubjub.py]: why is this async?

File: honeybadgermpc/progs/jubjub.py [hunk 7]
```diff
@@ -198,15 +194,24 @@ def __str__(self):
...
+        if not isinstance(other, SharedIdeal):
+            return False
+
+        return self.curve == other.curve
+
+    async def open(self):
```
Comment (smkuls) [inline honeybadgermpc/progs/jubjub.py]: Same here

File: tests/progs/test_jubjub.py [hunk 8]
```diff
@@ -84,24 +80,19 @@ def test_basic_point_functionality():
...
-        p3 = await SharedPoint.create(context, context.Share(
-            0), context.Share(1), Jubjub(Jubjub.Field(-2)))
 
-        assert await shared_point_equals(p1, p1)
-        assert not await shared_point_equals(p1, p2)
-        assert not await shared_point_equals(p1, p3)
-
-    await run_test_program(_prog, test_runner)
+        # Different curve
+        p3 = deepcopy(p1)
```
Comment (smkuls) [inline tests/progs/test_jubjub.py]: A deepcopy would be really slow, why do you need this? Can you replace this with something else?

**Review by smkuls — CHANGES_REQUESTED** [review]
I am not aware of the JubJub and MIMC code, @hanyunxu would be the best person to review that (looks like she has already approved these changes)

File: honeybadgermpc/mpc.py (hunk 1 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/mpc.py:89]: The asyncio documentation specifies that [`create_task`](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task) is preferred over `ensure_future` in python3.7. Furthermore, we mostly already mostly used `create_task`, so this makes it consistent throughout. 

When we call `ensure_future` with a coroutine,[ it simply calls `create_task` anyways](https://github.com/python/cpython/blob/56065d4c8ac03042cb7e29ffda9b1ac544a37b4d/Lib/asyncio/tasks.py#L616).

File: honeybadgermpc/mpc.py (hunk 2 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/mpc.py:68]: Currently, shares are already assigned ids in this order-- we get the length of `self._openings` and assign `share_id` that way, then assign an opening to that slot of `self._openings` before ever yielding execution to another task. This simply adds an extra layer of overhead to understanding the code given that we use `self._openings` for no other reason today.

Conversely,  the new code makes the assignment of ids much more explicit, and also saves the memory that `self._openings` took up. We never deleted entries once they finished opening, and never actually use `self._openings` other than to assign these IDs.

 I think in the current state of `Mpc`, this change is better.

Comment (smkuls) [reply honeybadgermpc/mpc.py:68]: Sounds good!

File: honeybadgermpc/mpc.py (hunk 3 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/mpc.py]: The second return value is actually `_confirmed_errors` from reconstruction when successful. I don't think this is worth logging here, but this makes it clearer in the code that the function is actually returning two things. 

That said, there is some error handling to be had here-- this returns `None` in the case of failed reconstruction. We should throw an exception here.

Comment (smkuls) [reply honeybadgermpc/mpc.py]: Do we add a log when there is an error? If not, then can you please add a log?

File: honeybadgermpc/mpc.py (hunk 4 above)
Comment (smkuls) [reply honeybadgermpc/mpc.py]: Shouldn't this return a `Future` which resolves to the opened values later on - as per our data flow?

Comment (Drake-Eidukas) [reply honeybadgermpc/mpc.py]: It keeps this in line with `open`. It was weird how one returned a future but the other didn't. 

Perhaps returning a `GFElementFuture` from `open` and a list of `GFElementFuture`s for `open_share_array` and making both of these non-async would be a better idea.

File: tests/progs/test_jubjub.py (hunk 8 above)
Comment (Drake-Eidukas) [reply tests/progs/test_jubjub.py]: I just replaced it with `copy`, which executes much faster.

File: honeybadgermpc/progs/jubjub.py (hunk 6 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/progs/jubjub.py]: It seems cleaner than returning a `Future` which will evaluate to `True` or `False`, as it's nonsensical to call either this or `open` without awaiting immediately (or gathering).

File: honeybadgermpc/progs/jubjub.py (hunk 7 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/progs/jubjub.py]: See prior comment, this also is to match style of `Share` objects today.

File: honeybadgermpc/mpc.py (hunk 5 above)
Comment (Drake-Eidukas) [reply honeybadgermpc/mpc.py]: With the current `_recvloop`, no, but I think it's best to have this in here for completeness.

File: honeybadgermpc/progs/jubjub.py (hunk 6 above)
Comment (smkuls) [reply honeybadgermpc/progs/jubjub.py]: I think I should have phrased my concern better, there is no `await` in this method and the one below, so why do you even need the `async` keyword?

Comment (Drake-Eidukas) [reply honeybadgermpc/progs/jubjub.py]: I think the semantics of SharedPoint and SharedIdeal should be the same, having this be async makes the code itself cleaner. Perhaps returning a future is better here.

**Comment by Drake-Eidukas** [comment]
notable changes-- 
- `Mpc`'s `open_share`, `open_share_array` methods, as well as `SharedPoint`'s `open` and `equals` methods now return futures and are not async
- spacing edits that I found helpful when reading through the failing `test_hbavss_light_parallel_share_array_open`. 

Also, it seems that `test_hbvavss_light_parallel_share_array_open` is failing every other time that I run this due to a reconstruction error, but this persists even on `dev`, so I'm chalking it up to something other than these changes.

**Review by smkuls — APPROVED** [review]
